EventListener/LeadTimelineSubscriber.php

```diff
@@ -0,0 +1,49 @@
+<?php
+
+namespace MauticPlugin\MauticContactSourceBundle\EventListener;
+
+use Mautic\CoreBundle\EventListener\CommonSubscriber;
+use Mautic\LeadBundle\Event\LeadTimelineEvent;
+use Mautic\LeadBundle\LeadEvents;
+
+class LeadTimelineSubscriber extends CommonSubscriber
+{
+    /**
+     * {@inheritdoc}
+     */
+    public static function getSubscribedEvents()
+    {
+        return [
+            LeadEvents::TIMELINE_ON_GENERATE => ['onTimelineGenerate', 0],
+        ];
+    }
+
+    /**
+     * @param LeadTimelineEvent $event
+     */
+    public function onTimelineGenerate(LeadTimelineEvent $event)
+    {
+        $repo         = $this->em->getRepository('MauticContactSourceBundle:Event');
+        $sourceEvents = $repo->getEventsByContactId($event->getLeadId());
+
+        if (!is_array($sourceEvents)) {
+            return;
+        }
+
+        foreach ($sourceEvents as $srcEvent) {
+            $srcEvent['eventLabel'] = [
+                'label' => 'Contact Source: '.$srcEvent['sourceName'],
+                'href'  => "/s/contactsource/view/{$srcEvent['contactsource_id']}",
+            ];
+            $srcEvent['event']     = '';
+            $srcEvent['eventType'] = ucfirst($srcEvent['type']);
```

Comment:
seems trivial, but I would have made a private method for any kind of String Transformaqtions (ucfirst) etc...

Reply:
Probably could have been done in the actual template, but in this case I'm ok with the result for now.